### DISCUSSION TITLE
2D telemac output

### DISCUSSION THREAD

**Discussion by eroome** [OpenDrift/opendrift#1293] [General]
Hi, 
Does opendrift work with a 2D telemac (selafin) output? 

cheers,
Ed

**Comment by knutfrode** [comment]
Hi,

There is a reader for Telemac/Selafin, which I believe was originally made by @Boorhin 
https://github.com/OpenDrift/opendrift/blob/74f61eb68830ca91cca5b901796112ae1ac16824/opendrift/readers/reader_telemac_selafin.py

There is a example script in the gallery, but the output is not shown, as it needs a Selafin package which is not included in OpenDrift distribution:
https://opendrift.github.io/gallery/example_long_selafin.html#sphx-glr-gallery-example-long-selafin-py

I dont know if this reader will work with 2D data, but it can proably be adapted to work.

**Reply by eroome** [reply]
Hi,

Thanks for the quick reply, I have been working on adapting the reader today but I am new to the structure of opendirft and python in general. Do you have any pointers as to what would need changing, I assume its a case of removing the depth related varaibles?

cheers,
Ed

**Reply by eroome** [reply]
Hi @knutfrode @Boorhin ,

Have either of you had recent success running the example script and reading the Telemac-3D file. 
I can't get it to read the seawater_velocity data.

Maybe it is a new issue which has arisen due to a python update or something?

cheers,
Ed

**Comment by Boorhin** [comment]
Hi,
The reader use the script from telemac (that is easy to replicate to be
honest). The number of layers is of course read so that is not an issue. If
I have an example of a 2D file I could check for compatibility but I don't
think it would be an issue. Did you try it?
Cheers

**Reply by eroome** [reply]
Hi @Boorhin ,

I did check for compatibility by adapting your example script (shown below). However, OceanDrift runs but it fails to read any values for x_seawater_velocity and y_seawater_velocity. Is your script designed to be compatible with 2D files?

from datetime import timedelta, datetime
from os import sep
from pyproj import Proj
from opendrift.readers import reader_telemac_selafin
from opendrift.models.oceandrift import OceanDrift

o = OceanDrift(loglevel=0)

filename='C:\\Users\edroo\Desktop\PhillipinesAtmosTest_xy.slf'

proj="+proj=utm +zone=51 +ellps=WGS84 +datum=WGS84 +units=m +no_defs"

        
start_time= datetime(2021,11,25,00,00)
selafin = reader_telemac_selafin.Reader(filename=filename,proj4 = proj, start_time=start_time)
o.add_reader(selafin)

o.set_config('general:coastline_action', 'previous')

time_step = timedelta(hours=0.5)
length= timedelta(hours=(35*24))
num_steps = ((35*48)+1)


lon,lat = 123.7716, 10.2259
o.seed_elements(lon=lon, lat=lat, radius=20, number= 200, \
   time= start_time)
o.run(time_step=timedelta(hours=0.5), duration=timedelta(1/24))

**Comment by Boorhin** [comment]
I cannot see an error message so it is difficult to judge. You can probe
the variables in your file with the reader.
Did you use the telemac script to see what's in your file first?
Cheers

**Comment by eroome** [comment]
Yes, I checked that, the variables are being read correctly by the reader. The error arises in the later part, when ocean drift needs to interpolate the velocity values from the selafin file. The error message is as follows...

argmin of an empty sequence
Traceback (most recent call last):
  File "C:\Users\edroo\opendrift\opendrift\models\basemodel\environment.py", line 646, in get_environment
    reader.get_variables_interpolated(
  File "C:\Users\edroo\opendrift\opendrift\readers\basereader\variables.py", line 898, in get_variables_interpolated
    env, env_profiles = self.get_variables_interpolated_xy(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\edroo\opendrift\opendrift\readers\basereader\variables.py", line 764, in get_variables_interpolated_xy
    env, env_profiles = self._get_variables_interpolated_(
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\edroo\opendrift\opendrift\readers\basereader\unstructured.py", line 61, in _get_variables_interpolated_
    env = self.get_variables(variables, time, reader_x, reader_y, z)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\edroo\opendrift\opendrift\readers\reader_telemac_selafin.py", line 231, in get_variables
    idx_layer = np.abs(pm - z).argmin(axis=0)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: attempt to get argmin of an empty sequence

**Comment by Boorhin** [comment]
Is it possible to have an extract of the file?
Not sure what the empty sequence means. Sounds like new python jargon.
I can adapt the reader to the 2D example if that's the issue.
Personally, I don't really like slf files, they are from another century ;)
I transform them to xarray zarr archives. It is easier to understand what's
in it. I have a repo with a script to do that if you are interested.
I still don't really know what's in your file, all kinds of things can be
stored in these Fortran files and I have had access to only few of them.
Cheers

**Reply by eroome** [reply]
Hi ,Yes that would be great.
Sorry, but I am not sure how to send an extract. I will run the model for short time period and send you the output.
I am also working on setting up a 3D model run, I can then test that output.

**Reply by eroome** [reply]
I just tried with this 3D output (download link below) and I get the same error - 

https://bangoroffice365-my.sharepoint.com/:u:/g/personal/dwr19fbr_bangor_ac_uk/EfZHeQlI6h5IiXyjbr8_EdEBxfF838E0ZIecXjiQ2eN_iQ?e=zZgISc

**Reply by eroome** [reply]
Here is an example 2D output 
https://bangoroffice365-my.sharepoint.com/:u:/g/personal/dwr19fbr_bangor_ac_uk/ERMfqNZim1ZBsnXYwFXHUt8B16Wc_5hsUdkgRZxk62MS-g?e=TsfZCn

**Reply by eroome** [reply]
Hi @Boorhin , 

Just checking over your 3D example file and I noticed that it only has the XY velocities in tho output. Was the script designed to work only with this type of output?

thanks for you help on this,
Ed

**Reply by Boorhin** [reply]
I will have a look

**Reply by Boorhin** [reply]
OK, the problem is that you have an old version of the reader and I never manage to pass the CI-CC established by the core team of OD and I honestly can't find the time to do it. 
Could you try with this reader?
https://github.com/Boorhin/opendrift/blob/master/opendrift/readers/reader_telemac_selafin.py
But one issue is the need of the variable 'ELEVATION Z     ' to make a variable 'sea_floor_depth_below_sea_level' for Opendrift.

**Reply by eroome** [reply]
I get the same error code with that reader, they look very similar. 
Does that reader work for you (with the example)?

**Reply by Boorhin** [reply]
So I investigated your 3D example, it does not contain an elevation variable (Z) so the reader will not manage to index the layers. What version of Telemac are you using?
This is using the telemac python interface, notice that you only have the velocities but no topology. 
![image](https://github.com/OpenDrift/opendrift/assets/9576982/d3f8977b-0ca3-45b7-abac-bda661cd854f)
It could be that they changed the way the Fortran files are coded but I really doubt it. So maybe it is a matter of exporting Z ? The 3D files I have contain much more variables including 'ELEVATION Z     '

**Reply by eroome** [reply]
I am using a very recent release TELEMAC/v8p5r0-issue981. I need to use this because it has a new atmospheric forcing method.
In my last post I was reffering to you example (https://opendrift.github.io/gallery/example_long_selafin.html) - this still does not work for me.

In reader_telemac_selafin, I change the path to this....  pytel = os.path.join('C:\\Users\edroo\opendrift', 'scripts', 'python3') --  (in line 29)
This shouldn't cause any issues right?

**Reply by eroome** [reply]
I will re-run my example case now - outputting Z.

**Reply by Boorhin** [reply]
I went through the 2D file and it is possible to make some reader out of it if I think about it. Should be relatively easy,
For the 3D file I see that there are 2 layers but I am unable to place them in space with what's in the file.

**Comment by eroome** [comment]
Hi Julien,

I no longer need a 2D file reader. From now onwards I will just run my model in 3D. But I am still having trouble running your example with the 3D reader.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by Boorhin** [comment]
> I am using a very recent release TELEMAC/v8p5r0-issue981. I need to use this because it has a new atmospheric forcing method. In my last post I was reffering to you example (https://opendrift.github.io/gallery/example_long_selafin.html) - this still does not work for me.
> 
> In reader_telemac_selafin, I change the path to this.... pytel = os.path.join('C:\Users\edroo\opendrift', 'scripts', 'python3') -- (in line 29) This shouldn't cause any issues right?

Hi, you have a configuration problem, you should not specifically change this line, you should create an environmental variable called HOMETEL as per specification of Telemac installation. In your case it should be 'C:\Users\edroo\opendrift'
Then pytel will work and thus the reader
Please tell me when you have a file, I will then try to see if there are more issues
cheers

**Reply by eroome** [reply]
So I made the HOMETEL variable and outputted 'Z' - the reader is getting x vel and y vel values now, thanks !

**Reply by Boorhin** [reply]
Glad it works!
in my to-do list I need to make a test if HOMETEL exists.

**Comment by eroome** [comment]
Yep, it works fine so far!

Thinking about it, I think I will need a 2D reader to get variables such as Depth, Wind X Y and Stokes Drift  from my 2D mesh.

I think there could be a easier work around though by using this xarray-selafin · PyPI<https://pypi.org/project/xarray-selafin/>

But my python skills are not good enough to clearly see how you can input the xarray dataset into open drift.

Ed
________________________________

**Reply by Boorhin** [reply]
yes I did something similar a long time ago
https://github.com/Boorhin/Telemac3D_to_xarray
But you can map any variable with the current reader, if you give me a list of what you want, it is not complicated to do
Again an example file is the easiest, I did based on the Telemac doc but they don't list the variables properly so I don't have an exhaustive list of all the potential variables. I then have to convert them into something that OD can read aka CF-compliant

**Reply by Boorhin** [reply]
My fork of opendrift contains a reader for xarray that never made it to the master branch of Opendrift. I am a bit tired of fighting the CI-CC system of the devs so I gave up but you can try it.
https://github.com/Boorhin/opendrift/blob/master/opendrift/readers/reader_xarray_CF_unstructured.py

**Comment by knutfrode** [comment]
Probably better than making a specific reader for Xarray, is to update the generic unstructured reader to accept an Xarray dataset, as alternative to providing filename or URL.
This approach is already used for  the generic structured reader and the ROMS reader.
https://github.com/OpenDrift/opendrift/blob/master/opendrift/readers/reader_netCDF_CF_generic.py
https://github.com/OpenDrift/opendrift/blob/master/opendrift/readers/reader_ROMS_native.py

However, this would require that the Telemac data structure is the same as expected in the generic unstructured reader, which I guess is to some degree adapted to FVCOM:
https://github.com/OpenDrift/opendrift/blob/master/opendrift/readers/reader_netCDF_CF_unstructured.py
